BugFix #14049: Can't filter on some date formats in ES

varunsharma27 opened the pull request:
Would've added a few tests to prevent regression but there were no existing tests for table.js
#14049: using value from _source to generate filters.
**Analysis:**

This bug exists since forever (Probably introduces in 5.x due to radical changes), let me give you steps to reproduce:

1. Set index mapping in dev-tools:
```
PUT company
{
  "mappings": {
    "properties": {
      "@timestamp": {
        "type": "date"
      },
      "founded_date": {
        "type": "date",
        "format": "dd/MM/yyyy"
      }
    }
  }
}
```
2. Index a document:
```
POST company/_doc
{
  "founded_date": "31/05/2014",
  "@timestamp": "2019-08-15T15:50:21.112Z"
}
```
3. Set index pattern.
4. Go to discover and add a filter on `founded_date` using the '+'.
![image](https://user-images.githubusercontent.com/18154526/63133581-d54f9f00-bfbd-11e9-9794-6f5d1ed9ab78.png)

![image](https://user-images.githubusercontent.com/18154526/63133590-e00a3400-bfbd-11e9-91c2-009dd4b10767.png)

This is mainly happening because kibana calls date fields with special formatting using `docvalue_fields` and the value from `_source` is overwritten in flatten when we decorate it with computed fields here: https://github.com/elastic/kibana/blob/master/src/legacy/ui/public/index_patterns/_flatten_hit.js#L77

And unfortunately this value is being used as it is to add a filter: https://github.com/elastic/kibana/blob/662a5abb9fdd580a2c91e974fee547f030062f12/src/legacy/core_plugins/kbn_doc_views/public/views/table.html#L8
This is the filter method which is being passed down: [filterQuery](https://github.com/elastic/kibana/blob/662a5abb9fdd580a2c91e974fee547f030062f12/src/legacy/core_plugins/kibana/public/discover/controllers/discover.js#L882)

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Pinging @elastic/kibana-app

nickpeihl commented:
jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13933/)

timroes commented:
Hi @varunsharma27,

thanks for opening this PR. Could you please merge master into your branch again, since there are currently some conflicts.

Thanks

@kertal @Bargs my RNG determined you as the designated reviewers for this ;-)

kertal commented:
@timroes your RNG implemented pretty well. 
@varunsharma27 note that the table component was converted to react/typescript in the meantime, so those conflicting files don't exist anymore in master. here are the new files that are relevant for this PR:

https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/kbn_doc_views/public/views/table/table.tsx
https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/kbn_doc_views/public/views/table/table_row.tsx

varunsharma27 commented on src/legacy/core_plugins/kbn_doc_views/public/views/table/table.tsx:
```diff
@@ -47,7 +48,7 @@ export function DocViewTable({
           .sort()
           .map(field => {
-            const valueRaw = flattened[field];
+            const valueRaw = get(hit._source, field, flattened[field]);
```
Get the value from `_source` if available else fallback to `flattened[field]` which would be mostly the case for meta fields such as `_index`, etc.

varunsharma27 commented:
@kertal, updated the branch with master and re-implemented the fix. Also added to test cases to prevent regression.
I found another peculiar behavior in kibana, so if you go and manually try to add this filter (Or add/edit any filter with such issue), you get:
![image](https://user-images.githubusercontent.com/18154526/64055975-1b912a80-cb88-11e9-8a22-ee025dcb68f2.png)

In case of trying to edit a filter added after this fix:
![image](https://user-images.githubusercontent.com/18154526/64055868-30b98980-cb87-11e9-871f-67253494dad6.png)
This is considered invalid due to this code: https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/data/public/filter/filter_bar/filter_editor/lib/filter_editor_utils.ts#L88
So as you can see, we are using moment here for date fields but that's not how Elasticsearch works. Due to this, even if a user wants to manually add the correct filter, it would be impossible as save it as kibana would consider the "correct" value to be invalid.
So as of now, there is no way in master to add such filters. With this fix, you may use the filter buttons and #14049 gets fixed (Which has been in kibana since 5.x and should be backported to currently supported versions).

However, I'd suggest you to open an issue for the above mentioned behavior.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/512/)

varunsharma27 commented:
@Bargs You are right, how about I:

- Introduce date format info to Index Patterns. This will also require migrations but kibana migrations are slightly basic and does not allow calls to ES (As far as I know). There can be other alternatives like firing another query to fetch date mapping info but they'd be very ugly and not optimal.
- Use that format in [filter editor utils](https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/data/public/filter/filter_bar/filter_editor/lib/filter_editor_utils.ts#L88) to validate date fields.
- While adding filters, only use `_source` if it's a date field and does not use default (or `strict_date_optional_time`) formatting. 

So as you can see, main problems arise when adding explicit match filters on dates where we might get wrong results or the Ui might block us from entering the 'correct' value. This makes it critical for us to have date format info from the mapping of index patterns so that we can adapt accordingly.

How do you think I should go about in this PR, or is it something you're not comfortable in getting PRs for at the moment?

Bargs commented:
In theory I think that's the right approach, but it will be quite difficult to support [every single format](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html) that ES supports.

There's also a pretty simple workaround to the original issue that this PR tries to fix. In the example from the original comment, you should be able to switch the format of `founded_date` from `dd/MM/yyyy` to `dd/MM/yyyy||strict_date_optional_time||epoch_millis`. This format would get filtering working without having to change the format of the source data.

So IMO a proper fix would be quite a lot of work for not much reward.

varunsharma27 commented:
@Bargs You might be right, and most users would be fine with the workaround. Would you suggest adding this date format issue to the [known limitations](https://www.elastic.co/guide/en/kibana/current/limitations.html) where we can explain the workaround and if enough business requirement arises, then it might become justified to rectify this limitation. How would you frame the wordings?

- Filters cause exception on unsupported (by kibana) date mapping.

And based on the current discussion, should #44557 be closed? As we still do not support epoch millis in the filter.

Feel free to close this PR or pick it up if/when the need arises.

Bargs commented:
I think adding this to the known limitations page is a great idea. I would just create a section called "Date formats" and then in the Date formats page you can list all the format related limitations. If we add it to the known limitations I think we could close #44557. Thanks for attempting to take this on, sorry it ended up being more tricky than expected!
